{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a30c6469",
   "metadata": {},
   "source": [
    "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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "379c8b05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: OPEN_AI_KEY=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\n"
     ]
    }
   ],
   "source": [
    "%env OPEN_AI_KEY = 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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8d5965a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: openai in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (0.28.1)\n",
      "Requirement already satisfied: wandb in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (0.15.12)\n",
      "Requirement already satisfied: requests>=2.20 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from openai) (2.31.0)\n",
      "Requirement already satisfied: tqdm in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from openai) (4.66.1)\n",
      "Requirement already satisfied: aiohttp in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from openai) (3.8.6)\n",
      "Requirement already satisfied: Click!=8.0.0,>=7.1 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (8.1.7)\n",
      "Requirement already satisfied: GitPython!=3.1.29,>=1.0.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (3.1.37)\n",
      "Requirement already satisfied: psutil>=5.0.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (5.9.0)\n",
      "Requirement already satisfied: sentry-sdk>=1.0.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (1.32.0)\n",
      "Requirement already satisfied: docker-pycreds>=0.4.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (0.4.0)\n",
      "Requirement already satisfied: PyYAML in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (6.0)\n",
      "Requirement already satisfied: pathtools in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (0.1.2)\n",
      "Requirement already satisfied: setproctitle in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (1.3.3)\n",
      "Requirement already satisfied: setuptools in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (68.0.0)\n",
      "Requirement already satisfied: appdirs>=1.4.3 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (1.4.4)\n",
      "Requirement already satisfied: protobuf!=4.21.0,<5,>=3.19.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from wandb) (4.24.4)\n",
      "Requirement already satisfied: six>=1.4.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from docker-pycreds>=0.4.0->wandb) (1.16.0)\n",
      "Requirement already satisfied: gitdb<5,>=4.0.1 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from GitPython!=3.1.29,>=1.0.0->wandb) (4.0.10)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from requests>=2.20->openai) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from requests>=2.20->openai) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from requests>=2.20->openai) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from requests>=2.20->openai) (2023.7.22)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from aiohttp->openai) (23.1.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from aiohttp->openai) (6.0.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from aiohttp->openai) (4.0.3)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from aiohttp->openai) (1.9.2)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from aiohttp->openai) (1.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from aiohttp->openai) (1.3.1)\n",
      "Requirement already satisfied: smmap<6,>=3.0.1 in /home/basticool/miniconda3/envs/bastienv/lib/python3.11/site-packages (from gitdb<5,>=4.0.1->GitPython!=3.1.29,>=1.0.0->wandb) (5.0.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install --upgrade openai wandb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5aaedccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import openai\n",
    "import wandb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "332593dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Tracking run with wandb version 0.15.12"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Run data is saved locally in <code>/home/basticool/Documents/projects/MiniMeditron-Prototype/data/structure/wandb/run-20231017_163049-gdb6tzzn</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Syncing run <strong><a href='https://wandb.ai/bastiteam/GPT-4%20in%20Python/runs/gdb6tzzn' target=\"_blank\">icy-pine-9</a></strong> to <a href='https://wandb.ai/bastiteam/GPT-4%20in%20Python' target=\"_blank\">Weights & Biases</a> (<a href='https://wandb.me/run' target=\"_blank\">docs</a>)<br/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View project at <a href='https://wandb.ai/bastiteam/GPT-4%20in%20Python' target=\"_blank\">https://wandb.ai/bastiteam/GPT-4%20in%20Python</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run at <a href='https://wandb.ai/bastiteam/GPT-4%20in%20Python/runs/gdb6tzzn' target=\"_blank\">https://wandb.ai/bastiteam/GPT-4%20in%20Python/runs/gdb6tzzn</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "run = wandb.init(project='GPT-4 in Python')\n",
    "prediction_table = wandb.Table(columns=[\"prompt\", \"prompt tokens\", \"completion\", \"completion tokens\", \"model\", \"total tokens\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "81ffb30a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nVanillyl mandelic acid\\n\\n# Overview\\nVanillyl mandelic acid (VMA) is a metabolite of the catecholamine: norepinephrine.\\nVMA is found in the urine, along with other catecholamine metabolites, including homovanillic acid (HVA).  In timed urine tests the quantity (concentration μg /24 h) is assessed, along with creatinine clearance, and the concentration of cortisols, catecholamines, and metanephrines.\\nThese urinalysis tests are used to diagnose an adrenal gland tumor called pheochromocytoma, a tumor of catecholamine-secreting chromaffin cells. These tests may also be used to diagnose neuroblastomas, and to monitor treatment of these conditions.\\nNorepinephrine breaks down into normetanephrine and VMA.  Norepinephrine is one of the hormones produced by the adrenal glands, which are found on top of the kidneys. They are released into the blood during times of physical or emotional stress, which are factors that may skew the results of the test.\\nde:Vanillinmandelsäure\", 'Vanillyl mandelic acid'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nSLC7A9\\nSolute carrier family 7 (cationic amino acid transporter, y+ system), member 9 also known as SLC7A9 is a protein which in humans is encoded by the SLC7A9 gene.\\n\\n# Function\\nThis gene encodes a protein that belongs to a family of light subunits of amino acid transporters. This protein plays a role in the high-affinity and sodium-independent transport of cystine and neutral and dibasic amino acids, and appears to function in the reabsorption of cystine in the kidney tubule. The protein associates with the protein coded for by SLC3A1.\\n\\n# Clinical significance\\nMutations in this gene cause non-type I cystinuria, a disease that leads to cystine stones in the urinary system due to impaired transport of cystine and dibasic amino acids.\", 'SLC7A9'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nTheria\\nTheria (Template:PronEng, from the Greek θηρίον, wild beast) is a subclass of mammals that give birth to live young without using a shelled egg, including both eutherians (placental mammals) and metatherians (marsupials and their ancestors).\\n\\n# Extent\\nThe subclass includes humans. They have external ears, most can suckle on a nipple, and have an ankle specialized for power and range of motion. Therians are often classified by their specialized dentition.\\nAlmost all currently extant (not extinct) mammals are therians. The only exceptions are the platypus and the echidnas (spiny anteater), both of which are prototherian monotremes.\\n\\n# Bibliography\\n- Vaughan, Terry A., James M. Ryan, and Nicholas J. Czaplewski. 2000. Mammalogy: Fourth Edition. Saunders College Publishing, 565 pp. ISBN 0-03-025034-X\", 'Theria'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nTOLLIP\\nToll interacting protein, also known as TOLLIP, is an inhibitory adaptor protein that in humans is encoded by the TOLLIP gene.\\n\\n# Function\\nIt is an inhibitory adaptor protein within Toll-like receptors (TLR). The TLR pathway is a part of the innate immune system that recognizes structurally conserved molecular patterns of microbial pathogens, leading to an inflammatory immune response.\\n\\n# Clinical significance\\nPolymorphisms in TLR genes have been implicated in various diseases like atopic dermatitis. Recently, variations in the TOLLIP gene have been associated with tuberculosis and idiopathic pulmonary fibrosis.\\n\\n# Interactions\\nTOLLIP has been shown to interact with TOM1, TLR 2, TLR 4 and IL1RAP.\", 'TOLLIP'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nCalmurid\\nCalmurid is the name of a cream that is manufactured by Galderma. Calmurid Cream contains the active ingredients lactic acid and urea, whereas Calmurid HC contains an additional ingredient, the mild corticosteroid hydrocortisone.\\nDue to lactic acid's keratolytic properties (to break down hard skin cells) and urea's hydrating properties, Calmurid is used primarily in the treatment of dry, scaly skin. Ichthyosis and general dermatitis in the absence of inflammation are some of its indications.\\nWhen the extra steroid component is added (as in Calmurid HC), it is used to treat dry, scaly skin that is accompanied by inflammation of the skin. This may include various forms of eczema. The presence of a corticosteroid, however, means that the cream should be used only sparingly and only for the shortest time period due to possible side effects that may occur due to systemic absorption of the steroid.\", 'Calmurid'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nSOX4\\nTranscription factor SOX-4 is a protein that in humans is encoded by the SOX4 gene.\\n\\n# Function\\nThis intronless gene encodes a member of the SOX (SRY-related HMG-box) family of transcription factors involved in the regulation of embryonic development and in the determination of the cell fate. The encoded protein may act as a transcriptional regulator after forming a protein complex with other proteins, such as syndecan binding protein (syntenin). The protein may function in the apoptosis pathway leading to cell death as well as to tumorigenesis and may mediate downstream effects of parathyroid hormone (PTH) and PTH-related protein (PTHrP) in bone development. The solution structure has been resolved for the HMG-box of a similar mouse protein.\\nSox4 is expressed in lymphocytes (B and T) and is required for B lymphocyte development.\\n\\n# Clinical significance\\nA genomic region close to the SOX4 gene has been associated with endometrial cancer development.\\n\\n# Interactions\\nSOX4 has been shown to interact with SDCBP.\", 'SOX4'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nAnthranilic acid\\nAnthranilic acid is the organic compound with the formula C6H4(NH2)CO2H.  This amino acid is  white solid when pure, although commercial samples may appear yellow.  The molecule consists of a benzene ring with two adjacent funcitonal groups, a carboxylic acid and an amine.  Because these two groups are polar, this organic compound is highly soluble in water. It is sometimes referred to as vitamin L\\nAnthranilic acid is biosynthesized from chorismic acid.  It is the precursor to the amino acid tryptophan via the attachment of a ribose to the amine group.\\n\\n# Uses\\nAnthranilic acid is used as an intermediate for production of dyes, pigments, and saccharin. It and its esters are used in preparing perfumes, pharmaceuticals and UV-absorber as well as corrosion inhibitors for metals and mold inhibitors in soya sauce.\\nAnthranilic acid can be used in organic synthesis to generate the benzyne intermediate.\", 'Anthranilic acid'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nLarsen syndrome\\nLarsen syndrome (LS), also called Autosomal Dominant Larsen syndrome, is a rare autosomal dominant congenital disorder that occurs in about every 1 in 100,000 people. Its symptoms include hypermobility, congenital dislocations, brachycephaly and cleft palate.\\nThe condition was first described in a 1950 journal report by L. J. Larsen, et al.\\n\\n# Genetics\\nLarsen syndrome is caused by mutations in the FLNB (Filamin B) gene.\\n\\n# Symptoms\\nA more complete list of symptoms includes:\\n- Multiple joint dislocations\\n- Foot deformities\\n- Non-tapering, cylindrical shaped fingers\\n- Unusual facial appearance\\n- Less commonly occurring:\\nShort stature\\nAdditional skeletal abnormalities\\nCleft palate\\nHeart defects\\nHearing impairment\\nMental retardation\\n- Short stature\\n- Additional skeletal abnormalities\\n- Cleft palate\\n- Heart defects\\n- Hearing impairment\\n- Mental retardation\", 'Larsen syndrome'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nPat Simmons\\nDr. Pat Simmons (born November 21, 1974 in Moose Jaw, Saskatchewan) is a Canadian curler. Simmons has skipped team Saskatchewan in four straight the Brier from 2005 to 2008.\\nSimmons made it to his first Brier in 2005 after winning his first provincial championship. He skipped Saskatchewan to a 6-5 finish, in a four-way tie for fifth place, out of the playoffs.\\nIn 2006, Simmons skipped Saskatchewan to a 5-6 finish at the Brier, out of the playoffs, tied for eighth place.\\nIn 2007, Simmons had a much better record, finishing 7-4, but narrowly missed the playoffs once again, finishing in fifth place.\\nIn 2008, Simmons finally skipped Saskatchewan to a playoff spot at the Brier, finishing the round robin in second place with a 9-2 record. However, he lost in the semi-final in an extra end to Ontario, skipped by Glenn Howard\", 'Pat Simmons'), (\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\nAmatoxin\\nAmatoxins are a subgroup of toxins found in Amanita phalloides and several other members of the genus Amanita, as well as some Conocybe, Galerina and Lepiota mushroom species.\\nTheir swift intestinal absorption coupled with their thermostability explains why their toxic effects occur in a relative short period of time. The most severe effects are toxic hepatitis with centrolobular necrosis and hepatic steatosis, as well as acute tubulointerstitial nephropathy, which altogether induce a severe hepatorenal syndrome (with a potential fatal outcome).\\nThe estimated minimum lethal dose is 0.1 mg/kg or 7 mg of toxin in adults. There are eight amatoxins:\\n- alpha-amanitin\\n- beta-amanitin\\n- gamma-amanitin\\n- epsilon-amanitin\\n- amanullin\\n- amanullinic acid\\n- proamanullin\\nda:Amatoxiner\\nit:Amatossine\\nfi:Amatoksiini\", 'Amatoxin')]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import random\n",
    "import json\n",
    "\n",
    "instruction = \"\"\"Using the provided medical guideline, craft a doctor's description of a patient. The description should detail symptoms from the guideline without explicitly naming the disease. You can enhance the authenticity of the description by incorporating the patient's pain levels related to symptoms, lifestyle, medical history, and any previously applied treatments. While you may introduce supplementary symptoms, they should be relevant and factual and the overall description should remain concise.\\n\"\"\"\n",
    "\n",
    "file_list = os.listdir(\"Guidelines/split_guidelines/wikidoc.jsonl\")\n",
    "\n",
    "num_smallest_files = int(len(file_list) * 0.1)\n",
    "\n",
    "sorted_files = sorted(file_list, key=lambda x: os.path.getsize(os.path.join(\"Guidelines/split_guidelines/wikidoc.jsonl\", x)))\n",
    "\n",
    "if num_smallest_files >= 10:\n",
    "    smallest_files = random.sample(sorted_files[:num_smallest_files], 10)\n",
    "else:\n",
    "    print(\"There are not enough small files to select 10 random files.\")\n",
    "\n",
    "prompt_tab = []\n",
    "\n",
    "for name in smallest_files:\n",
    "    with open(\"Guidelines/split_guidelines/wikidoc.jsonl/\" + name) as f:\n",
    "        #guideline = f.read()\n",
    "        jj = json.load(f)\n",
    "    prompt_tab += [(instruction + jj['text'], jj['title'])]\n",
    "\n",
    "print(prompt_tab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9e4c906e",
   "metadata": {},
   "outputs": [],
   "source": [
    "openai.api_key = os.getenv(\"OPEN_AI_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a485a347",
   "metadata": {},
   "outputs": [],
   "source": [
    "response_tab = []\n",
    "\n",
    "for (prompt, name) in prompt_tab:\n",
    "    message=[{\"role\": \"user\", \"content\": prompt}]\n",
    "    response = openai.ChatCompletion.create(\n",
    "        model=\"gpt-3.5-turbo\",\n",
    "        messages = message,\n",
    "        temperature=0.2,\n",
    "        max_tokens=1000,\n",
    "        frequency_penalty=0.0\n",
    "    )\n",
    "    response_tab += [(response, name)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "af1b7386",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the Generated_descriptions folder if it doesn't exist\n",
    "if not os.path.exists(\"Generated_descriptions\"):\n",
    "    os.makedirs(\"Generated_descriptions\")\n",
    "\n",
    "# from response_tab to save to file\n",
    "for (response, name) in response_tab:\n",
    "    with open(\"Generated_descriptions/\" + name + \".txt\", \"w\") as f:\n",
    "        f.write(response[\"choices\"][0][\"message\"][\"content\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bf79d5b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Waiting for W&B process to finish... <strong style=\"color:green\">(success).</strong>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run <strong style=\"color:#cdcd00\">icy-pine-9</strong> at: <a href='https://wandb.ai/bastiteam/GPT-4%20in%20Python/runs/gdb6tzzn' target=\"_blank\">https://wandb.ai/bastiteam/GPT-4%20in%20Python/runs/gdb6tzzn</a><br/> View job at <a href='https://wandb.ai/bastiteam/GPT-4%20in%20Python/jobs/QXJ0aWZhY3RDb2xsZWN0aW9uOjEwNzQxOTc5NQ==/version_details/v1' target=\"_blank\">https://wandb.ai/bastiteam/GPT-4%20in%20Python/jobs/QXJ0aWZhY3RDb2xsZWN0aW9uOjEwNzQxOTc5NQ==/version_details/v1</a><br/>Synced 6 W&B file(s), 1 media file(s), 1 artifact file(s) and 0 other file(s)"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Find logs at: <code>./wandb/run-20231017_163049-gdb6tzzn/logs</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "Error",
     "evalue": "You must call wandb.init() before wandb.log()",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mError\u001b[0m                                     Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[35], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m prediction_table \u001b[38;5;241m=\u001b[39m wandb\u001b[38;5;241m.\u001b[39mTable(columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPrompt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResponse\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTokens\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m      3\u001b[0m prediction_table\u001b[38;5;241m.\u001b[39madd_data(prompt,response[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchoices\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;241m0\u001b[39m],response[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124musage\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[0;32m----> 4\u001b[0m wandb\u001b[38;5;241m.\u001b[39mlog({\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpredictions\u001b[39m\u001b[38;5;124m'\u001b[39m: prediction_table})\n\u001b[1;32m      5\u001b[0m wandb\u001b[38;5;241m.\u001b[39mfinish()\n",
      "File \u001b[0;32m~/miniconda3/envs/bastienv/lib/python3.11/site-packages/wandb/sdk/lib/preinit.py:36\u001b[0m, in \u001b[0;36mPreInitCallable.<locals>.preinit_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpreinit_wrapper\u001b[39m(\u001b[38;5;241m*\u001b[39margs: Any, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Any:\n\u001b[0;32m---> 36\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m wandb\u001b[38;5;241m.\u001b[39mError(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou must call wandb.init() before \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m()\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mError\u001b[0m: You must call wandb.init() before wandb.log()"
     ]
    }
   ],
   "source": [
    "for prompt in prompt_tab:\n",
    "    prediction_table = wandb.Table(columns=[\"Prompt\", \"Response\", \"Tokens\"])\n",
    "    prediction_table.add_data(prompt,response['choices'][0],response['usage'])\n",
    "    wandb.log({'predictions': prediction_table})\n",
    "    wandb.finish()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cfab7a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the wandb storage\n",
    "wandb.restore('wandb/latest-run/files/predictions_table.jsonl', run_path='GPT-4 in Python/2j1q2q7i')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
